fix: Chat - Skeleton placeholder flickers when opening chat with IOU in offline mode

paultsimura opened the pull request:
### Details

Fix the flickering of the IOU report when offline

### Fixed Issues

$ https://github.com/Expensify/App/issues/30503
PROPOSAL: https://github.com/Expensify/App/issues/30503#issuecomment-1782978278


### Tests

Same as QA.

### Offline tests

Same as QA.

### QA Steps

**Precondition:** There must be a user with an existing unpaid IOU.

1. Log out and log in again and do not open the chat report with the user that has existing IOU.
2. Go offline.
3. Go to + > Request money > Manual.
4. Request money from the user in the precondition.
5. Click on the IOU report preview
- [ ] Verify that the skeleton UI is not flickering

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>


https://github.com/Expensify/App/assets/12595293/1accebf9-33ea-4a2e-ba57-0b0bad733765





</details>

<details>
<summary>Android: mWeb Chrome</summary>


https://github.com/Expensify/App/assets/12595293/fff2241f-c0da-4baa-a901-ed8c54901f65

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: Native</summary>



https://github.com/Expensify/App/assets/12595293/9fd6e8af-e19e-4874-a8eb-df9f3b381330





<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: mWeb Safari</summary>



https://github.com/Expensify/App/assets/12595293/b408da6b-77f8-428e-8af8-4251fddd313f





<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>


https://github.com/Expensify/App/assets/12595293/3fd9b821-b917-4507-aa96-4945a7d1ec2e



<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Desktop</summary>


https://github.com/Expensify/App/assets/12595293/e208cbe6-97b5-4145-89a6-c7b2701b24a3



<!-- add screenshots or videos here -->

</details>

paultsimura committed:
Prevent fetching older actions when offline

paultsimura committed:
Merge branch 'main' into fix/30503-report-skeleton

# Conflicts:
#	src/pages/home/report/ReportActionsView.js

paultsimura commented:
@getusha the PR's ready, the only thing I'm concerned about is that the skeleton UI is not visible on the first render of the IOU report. It appears only if we re-render it (e.g. go to transaction view and back).

But this seems to have been done intentionally:

> Skip this hook on the first render, as we are not sure if more actions are going to be loaded
> Therefore showing the skeleton on footer might be misleading

https://github.com/Expensify/App/blame/36495e609c2689742e5c7098cc8c0000c891e6fd/src/pages/home/report/ReportActionsList.js#L371-L376

getusha commented:
>the only thing I'm concerned about is that the skeleton UI is not visible on the first render of the IOU report.

@paultsimura is this happening on `main`?

aldo-expensify commented on src/pages/home/report/ReportActionsView.js:
```diff
@@ -78,13 +79,15 @@ function ReportActionsView(props) {
      * @returns {Boolean}
      */
     const isReportFullyVisible = useMemo(() => getIsReportFullyVisible(isFocused), [isFocused]);
+    const isFullHistoryFetched = useMemo(() => _.last(props.reportActions).actionName === CONST.REPORT.ACTIONS.TYPE.CREATED, [props.reportActions]);
```
I don't think `_.last(props.reportActions).actionName === CONST.REPORT.ACTIONS.TYPE.CREATED` is expensive enough to be using `useMemo`. This is just accessing the lsat member of an array which should be very fast.

paultsimura commented:
@ospfranco @fedirjh  could you please help us understand what use case was this piece of code initially fixing, and if it might be safe to remove now since we've separated the actions loading variables to `isLoadingNewerReportActions`, `isLoadingOlderReportActions`, and `isLoadingInitialReportActions`?

https://github.com/Expensify/App/blob/36495e609c2689742e5c7098cc8c0000c891e6fd/src/pages/home/report/ReportActionsList.js#L371-L376

paultsimura committed:
Show skeleton on first render when offline

paultsimura committed:
Revert forced offline olderReportActions skeleton UI

paultsimura committed:
Add missing dependencies

paultsimura commented:
@getusha I think I figured it out. We use the `useRef` here to not show the skeleton while the first request to load more actions is running (so that we won't show it just for a second):

https://github.com/Expensify/App/blob/c8e7ee5718730aef7eaeec052ee5c97e4284ff53/src/pages/home/report/ReportActionsList.js#L371-L376

But as we are offline, the actual request will not be sent to the BE, and we are safe to show the skeleton permanently until we are back online. So I've modified the logic with these references.

I also removed the originally proposed change to `isLoadingOlderReportActions` in `ReportActionsView` because when offline, we are already showing the `isLoadingInitialReportActions` skeleton, and I think this should be enough for the good UX (most likely, it was like this by design).

paultsimura committed:
Merge branch 'main' into fix/30503-report-skeleton

paultsimura commented:
> Jest Unit Tests are failing

@getusha I believe it was some flaky test. Resolved by itself after I merged `main`.

getusha commented:
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

</details>

getusha commented:
@paultsimura noticed this, still looks glitchy but with bigger interval than the previous one. can you check it?

https://github.com/Expensify/App/assets/75031127/42a1c2aa-3578-4449-b7e8-b77549c89aad

paultsimura commented:
@getusha Could you please share a little context on how to reproduce this? I see on your recording, this is happening on a regular report screen, not the IOU as it was in the issue. I'll try to reproduce, but any details would be useful 🙏🏼

getusha commented:
@paultsimura i just scrolled up, and this is happening. i think we should also fix this since it the same issue at a different place.

paultsimura commented:
> we should also fix this since it the same issue at a different place.

Sure, I'm on it. Thanks for catching this bug

paultsimura commented:
@getusha I couldn't reproduce it on either iOS or Android.
I see your recording is from Android – could you please double-check that you are testing after the app is already bundled, so it's not a cached version? Androids can get annoying by not loading the changes.

And also for the clarity of the experiment – please check if the same flickering is happening on the IOU Report

paultsimura commented:
@getusha I couldn't reproduce it on either iOS or Android.
I see your recording is from Android – could you please double-check that you are testing after the app is already bundled, so it's not a cached version? Androids can get annoying by not loading the changes.

And also for the clarity of the experiment – please check if the same flickering is happening on the IOU Report 🙏

getusha commented:
@paultsimura still not able to reproduce? i thought i saw a comment saying u're able to reproduce it.

paultsimura commented:
@getusha I tried on web - checked staging by mistake, and commented "reproduced". Only after that I noticed that it was staging, so I removed that comment.

Still not able to reproduce on my branch.

getusha reviewed (approved):
Looks good and tests went well.

getusha commented:
@paultsimura Looks like we have a conflict.

paultsimura committed:
Merge branch 'main' into fix/30503-report-skeleton

# Conflicts:
#	src/pages/home/report/ReportActionsView.js

paultsimura committed:
Use props.network instead of useNetwork

paultsimura commented:
@getusha thanks, resolved. Also made a tiny change in the last commit.

aldo-expensify commented:
Why did we change from using the hook `useNetwork` to passing the network as a prop here: https://github.com/Expensify/App/pull/30658/commits/772715011ce1431884bb499164faa36dcc8006e6

Are you sure that this won't cause issues? Are we going to need to retest this?

paultsimura commented:
@aldo-expensify the network prop has already been there (passed through `withNetwork()`) and was used in one place in this component:

https://github.com/Expensify/App/blob/d75501f485c7a8668f13eb09fca56e6281716335/src/pages/home/report/ReportActionsView.js#L124

I just noticed it when merging the latest `main`, so decided to not mix `withNetwork` and `useNetwork` in one component as they are interchangeable. Also, I did a retest, and it works as expected.

aldo-expensify commented:
I see now, thanks for explaining!

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/6795287634) to staging by https://github.com/aldo-expensify in version: 1.3.96-6 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/6818231417) to production by https://github.com/puneetlath in version: 1.3.96-15 🚀

platform | result
---|---
🤖 android 🤖|failure ❌
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/6826740157) to staging by https://github.com/aldo-expensify in version: 1.3.98-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/6869834824) to production by https://github.com/roryabraham in version: 1.3.98-5 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

s77rt commented on src/pages/home/report/ReportActionsView.js:
```diff
@@ -177,8 +177,8 @@ function ReportActionsView(props) {
-        // Only fetch more if we are not already fetching so that we don't initiate duplicate requests.
-        if (props.isLoadingOlderReportActions) {
+        // Only fetch more if we are neither already fetching (so that we don't initiate duplicate requests) nor offline.
+        if (props.network.isOffline || props.isLoadingOlderReportActions) {
```
This caused a regression https://github.com/Expensify/App/issues/31367. Previously even if we were offline  `Report.getOlderActions` would be still called and sets `isLoadingOlderReportActions` to `true` which in turns show the skeleton view. Now with this early return we prevented the skeleton view from showing up in this case.
